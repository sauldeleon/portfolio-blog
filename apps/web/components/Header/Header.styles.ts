import { styled } from 'styled-components'

import { CircleLink } from '@sdlgr/circle-link'

export const StyledCircleLink = styled(CircleLink)`
  display: none;

  ${({ theme }) => theme.media.up.md} {
    display: flex;
  }
`
