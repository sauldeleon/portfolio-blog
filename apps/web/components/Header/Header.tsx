import { Header as HeaderLib } from '@sdlgr/header'

import { useClientTranslation } from '@web/i18n/client'

export function Header() {
  const { t } = useClientTranslation('header')
  return (
    <HeaderLib
      items={[
        { href: '/experience', label: t('experience') },
        { href: '/contact', label: t('contact') },
        { href: '/portfolio', label: t('portfolio') },
      ]}
      actionButtonLabel={t('getInTouch')}
    />
  )
}
