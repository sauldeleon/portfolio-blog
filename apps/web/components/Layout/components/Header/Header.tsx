import { ArrowRightIcon } from '@sdlgr/assets'
import { Header as HeaderLib } from '@sdlgr/header'

import { useClientTranslation } from '@web/i18n/client'

import { StyledCircleLink, StyledLogoLink, StyledSLLogo } from './Header.styles'

export function Header() {
  const { t } = useClientTranslation('header')
  return (
    <HeaderLib
      aria-label={t('mainMenu')}
      logo={
        <StyledLogoLink href="/" aria-label={t('headerLogo')}>
          <StyledSLLogo height={55} width={106} />
        </StyledLogoLink>
      }
      navItems={[
        { href: '/experience', label: t('experience') },
        { href: '/contact', label: t('contact') },
        { href: '/portfolio', label: t('portfolio') },
        { href: '/portfolio', label: t('getInTouch'), hideOnDesktop: true },
      ]}
      actionItem={
        <StyledCircleLink
          href="/portfolio"
          label={t('getInTouch')}
          iconContent={<ArrowRightIcon />}
        />
      }
    />
  )
}
