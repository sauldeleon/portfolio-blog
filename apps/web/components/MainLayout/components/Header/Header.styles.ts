import styled from 'styled-components'

import { SLLogo } from '@sdlgr/assets'
import { CircleLink } from '@sdlgr/circle-link'
import { Link } from '@sdlgr/link'

export const StyledLogoLink = styled(Link)`
  display: flex;
  margin-right: 15px;
`

export const StyledSLLogo = styled(SLLogo)`
  color: ${({ theme }) => theme.colors.black};
  margin-bottom: 25px;

  ${({ theme }) => theme.media.up.md} {
    color: ${({ theme }) => theme.colors.white};
    margin-bottom: 0;
  }
`

export const StyledCircleLink = styled(CircleLink)`
  display: none;

  ${({ theme }) => theme.media.up.md} {
    margin: 0 20px;
    display: flex;
  }
`
