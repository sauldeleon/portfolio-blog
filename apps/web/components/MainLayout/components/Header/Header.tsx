import { usePathname, useSelectedLayoutSegments } from 'next/navigation'
import { useContext } from 'react'

import { ArrowRightIcon } from '@sdlgr/assets'
import { Header as HeaderLib } from '@sdlgr/header'
import { LanguageContext } from '@sdlgr/i18n-tools'

import { useClientTranslation } from '@web/i18n/client'

import { StyledCircleLink, StyledLogoLink, StyledSLLogo } from './Header.styles'

export function Header() {
  const { t } = useClientTranslation('header')
  const pathname = usePathname()
  const segments = useSelectedLayoutSegments()
  const { language } = useContext(LanguageContext)

  const isValidPath = !segments.some((segment) => segment.includes('/'))

  const navItems: {
    href: string
    label: string
    ariaLabel: string
    isActive: boolean
    hideOnDesktop?: boolean | undefined
  }[] = [
    {
      href: `/${language}/experience/`,
      label: t('experience'),
      ariaLabel: t('experienceAria'),
    },
    {
      href: `/${language}/contact/`,
      label: t('contact'),
      ariaLabel: t('contactAria'),
    },
    {
      href: `/${language}/portfolio/`,
      label: t('portfolio'),
      ariaLabel: t('portfolioAria'),
    },
    {
      href: `/${language}/blog/`,
      label: t('blog'),
      ariaLabel: t('blogAria'),
      hideOnDesktop: true,
      alternativeLinks: [new RegExp(`^/${language}/blog/`)],
    },
  ].map((navItem) => ({
    ...navItem,
    isActive:
      isValidPath &&
      (pathname?.startsWith(navItem.href) ||
        !!navItem.alternativeLinks?.some((regex) => regex.test(pathname))),
  }))

  return (
    <HeaderLib
      aria-label={t('mainMenu')}
      logo={
        <StyledLogoLink href={`/${language}/`} aria-label={t('headerLogo')}>
          <StyledSLLogo height={55} width={106} />
        </StyledLogoLink>
      }
      navItems={navItems}
      actionItem={
        <StyledCircleLink
          href={`/${language}/get-in-touch`}
          label={t('blog')}
          aria-label={t('blogAria')}
          iconContent={<ArrowRightIcon />}
        />
      }
    />
  )
}
