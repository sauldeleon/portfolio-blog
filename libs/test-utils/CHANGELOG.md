# Changelog Test utils

## [0.2.3](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.2...test-utils-0.2.3) (2024-01-08)

## [0.2.2](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.1...test-utils-0.2.2) (2023-09-26)

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.0...test-utils-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.1.0...test-utils-0.2.0) (2023-09-05)

### Bug Fixes

- reduced duplicated lines ([f51168a](https://github.com/sauldeleon/portfolio-blog/commit/f51168aefe025a5a65ffdb2dc22f7a16b863406e))
- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add contact initial page ([fea862b](https://github.com/sauldeleon/portfolio-blog/commit/fea862b3507268013149050ae3f05fa517e7b3d9))
- add homepage ([60523aa](https://github.com/sauldeleon/portfolio-blog/commit/60523aa31ec088d318745ef0c757b4cbe656f0d4))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
- toggle language button ([afd97e9](https://github.com/sauldeleon/portfolio-blog/commit/afd97e95beac0f70ed643c753e71db2ca014054b))
