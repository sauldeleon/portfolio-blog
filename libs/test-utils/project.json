{
  "name": "test-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Test utils \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore(${projectName}): release: v${version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/test-utils/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
