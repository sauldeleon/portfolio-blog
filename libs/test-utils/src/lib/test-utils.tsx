import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { Matcher } from '@testing-library/dom/types/matches'
import {
  RenderOptions,
  RenderResult,
  queryAllByAttribute as _queryAllByAttribute,
  queryByAttribute as _queryByAttribute,
  act,
  render,
  screen,
} from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import Router, { NextRouter } from 'next/router'
import { ReactNode, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { LanguageContextProvider } from '@sdlgr/i18n-config'
import { mainTheme } from '@sdlgr/main-theme'

export type MatcherCustomFnParams = {
  content: string
  element: Element | null
}

/* istanbul ignore next */
export const renderWithTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={mainTheme}>{children}</ThemeProvider>)
}

/* istanbul ignore next */
export function RenderProviders({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return (
    <QueryClientProvider client={queryClient}>
      <LanguageContextProvider value={{ language: 'en' }}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </LanguageContextProvider>
    </QueryClientProvider>
  )
}

/* istanbul ignore next */
export const renderApp = (
  children: ReactNode,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(<RenderProviders>{children}</RenderProviders>, options)
}

/**
 * Generic function to get element by attribute and value
 * @example queryByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function queryByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example queryByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function queryAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Generic function to get element by attribute and value
 * @example getByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function getByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example getByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function getAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Shortcut version for queryByAttribute, but specific for id
 * @param {HTMLElement} container
 * @param {string} id
 * @return {any}
 */
export function getById(container: HTMLElement, id: string) {
  return queryByAttribute('id')(container, id) as Element
}

/**
 * Shortcut version for queryByAttribute, but specific for class name
 * @param {HTMLElement} container
 * @param {string} className
 * @return {any}
 */
export function getByClassName(container: HTMLElement, className: string) {
  return queryByAttribute('class')(container, className) as Element
}

function matcherCustomFn(
  tag: string,
  matcher: Matcher,
  { element, content }: MatcherCustomFnParams
) {
  const isTag = element?.tagName.toUpperCase() === tag.toUpperCase()
  let matching = false
  if (isTag) {
    matching = content === matcher
    /* istanbul ignore else*/
    if (typeof matcher === 'function') {
      matching = matcher(content, element)
    } else if (matcher instanceof RegExp) {
      matching = matcher.test(content)
    }
  }
  return matching && isTag
}

/**
 * Get 1 element by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getByTagAndText(tag: string, matcher: Matcher) {
  return screen.getByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Get multiple elements by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getAllByTagAndText(tag: string, matcher: Matcher) {
  return screen.getAllByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Mock Math.random() to return a specific value
 * @param {number} value
 */
export function mockRandom(value = 0.1234) {
  const mockMath = Object.create(global.Math)
  mockMath.random = () => value
  global.Math = mockMath
}

/**
 * Hook to use in test to ensure "last" render is called
 */
export function useWithRefreshInTest() {
  const [withRefreshInTest, setWithRefreshInTest] = useState(false)

  useEffect(() => {
    // Change state to "rerender" the component
    setWithRefreshInTest(true)
  }, [])

  return withRefreshInTest
}

export { render } from '@testing-library/react'

/* istanbul ignore next */
export const awaitAllAsync = async () =>
  act(() => new Promise((resolve) => setTimeout(resolve, 0)))

/* istanbul ignore next */
export function renderWithNextRouter(
  children: ReactNode,
  options?: Partial<NextRouter>
) {
  return renderWithTheme(
    <RouterContext.Provider value={{ ...Router, ...options }}>
      {children}
    </RouterContext.Provider>
  )
}
