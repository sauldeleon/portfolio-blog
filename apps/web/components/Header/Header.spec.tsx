import { screen } from '@testing-library/react'

import { renderApp } from '@sdlgr/test-utils'

import { Header } from './Header'

describe('Header', () => {
  it('should render successfully', async () => {
    const { baseElement } = renderApp(<Header />)
    const nav = await screen.findByRole('navigation')
    expect(nav).toBeTruthy()
    expect(screen.getAllByRole('link')).toHaveLength(4)
    expect(baseElement).toMatchSnapshot()
  })
})
