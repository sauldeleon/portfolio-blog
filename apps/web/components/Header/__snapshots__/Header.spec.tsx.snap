// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render successfully 1`] = `
.c2 {
  font-family: 'Roboto mono',monospace;
  font-weight: 400;
  color: #FFFFFF;
  font-size: 0.9rem;
  line-height: normal;
  white-space: pre-line;
}

.c2 ul,
.c2 ol {
  margin-left: 40px;
  margin-bottom: 10px;
}

.c2 ul {
  list-style: outside disc;
}

.c2 ol {
  list-style: decimal;
}

.c2 li {
  padding-left: 10px;
}

.c3 {
  outline: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  max-height: 42px;
}

.c3:hover circle.loading {
  stroke: white;
  stroke-width: 6px;
  stroke-dasharray: 110;
  stroke-dashoffset: 82;
  animation: clock-loading 0.5s steps(4, jump-none) forwards;
  transform: rotate(-90deg);
  transform-origin: center;
}

.c4 {
  position: relative;
  width: 42px;
  height: 42px;
}

.c5 {
  position: absolute;
  transform: translate(-50%, -50%);
  top: 50%;
  left: 50%;
}

.c6 {
  position: absolute;
  transform: translate(-50%, -50%);
  top: 50%;
  left: 50%;
  display: flex;
}

.c7 {
  margin-left: 8px;
}

.c0 {
  height: 115px;
  max-width: 1440px;
  width: 100%;
  margin: 0 auto;
  display: flex;
  justify-items: center;
  align-items: center;
}

.c1 {
  display: flex;
  align-items: center;
  flex-grow: 1;
  justify-content: center;
  gap: 60px;
}

<body>
  <div>
    <nav
      class="c0"
    >
      <a
        href="/"
      >
        <svg
          height="55"
        >
          slLogo.svg
        </svg>
      </a>
      <ul
        class="c1"
      >
        <li
          class=""
        >
          <a
            href="/experience"
          >
            <p
              class="c2"
            >
              Experience
            </p>
          </a>
        </li>
        <li
          class=""
        >
          <a
            href="/contact"
          >
            <p
              class="c2"
            >
              Contact
            </p>
          </a>
        </li>
        <li
          class=""
        >
          <a
            href="/portfolio"
          >
            <p
              class="c2"
            >
              CV/Portfolio
            </p>
          </a>
        </li>
      </ul>
      <a
        class="c3"
        href="/"
      >
        <div
          class="c4"
        >
          <svg
            class="c5"
            height="42"
            width="42"
          >
            circleButton.svg
          </svg>
          <div
            class="c6"
          >
            <svg>
              arrowRight.svg
            </svg>
          </div>
        </div>
        <p
          class="c2 c7"
        >
          Get in touch
        </p>
      </a>
    </nav>
  </div>
</body>
`;
